opm/simulators/linalg/cuistl/detail/cusparse_matrix_operations.cu

```diff
@@ -0,0 +1,99 @@
+/*
+  Copyright 2022-2023 SINTEF AS
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <opm/simulators/linalg/cuistl/detail/cusparse_matrix_operations.hpp>
+#include <vector>
+#include <iostream>
+namespace Opm::cuistl::detail
+{
+
+namespace
+{
+    template <class T> __global__ void cuInvertDiagonalAndFlatten(T* matNonZeroValues, int rowIndices[], int colIndices[], size_t numberOfRows, size_t blocksize, T* vec)
+    {
+        const auto thrIndex = blockDim.x * blockIdx.x + threadIdx.x;
+
+        if (thrIndex < numberOfRows){
+            size_t nnzIdx = rowIndices[thrIndex];
+            size_t nnzIdxLim = rowIndices[thrIndex+1]; 
+
+            // this loop will cause some extra checks that we are within the limit in the case of the diagonal having a zero element
+            for (;colIndices[nnzIdx] != thrIndex && nnzIdx <= nnzIdxLim;) { 
+                nnzIdx++;
+            }
+
+            // pDiagBlock points to the start of where the diagonal block is stored
+            T* pDiagBlock = (matNonZeroValues+(blocksize*blocksize*nnzIdx));
+            // pVecBlock points to the start of the block element in the vector where the inverse of the diagonal block element should be stored
+            T* pVecBlock = (vec + (blocksize*blocksize*thrIndex));
+
+            // untested - based on Dune implementation
+            if (blocksize == 2){
```

Comment:
Would it be better to have the 2x2 and 3x3 cases as separate kernels, and dispatch among them in the invertDiagonalAndFlatten() function instead of the `if` on every thread? Also, then we can better handle the cases that will fail (blocksize == 4). What if blocksize is 1 btw? We should support that.

Similar arguments can be made for the elementWiseMultiplyMVKernel.
Similar arguments can be made about the vector_operation

Reply:
Not sure if the difference in practice really matters, but in case the compiler can do some neat trick with knowledge of the blocksize I have split it up now

Reply:
When writing the tests for the case handling blocksize=1 I realized why I had not implemented it already. CuSparse does not always support it, for instance not in the case of bsrmv, which is a requirement for jacobi preconditioning. We can of course implement our own sparse multiplication for this case if the simulator needs to support it.